mlir/tools/mlir-clang/Lib/clang-mlir.cc

```diff
@@ -1419,7 +1425,51 @@ ValueWithOffsets MLIRScanner::VisitReturnStmt(clang::ReturnStmt *stmt) {
   return nullptr;
 }
 
-std::map<const FunctionDecl *, mlir::FuncOp> functions;
+mlir::LLVM::LLVMFuncOp
+MLIRASTConsumer::GetOrCreateLLVMFunction(const FunctionDecl *FD) {
+  if (llvmFunctions.find(FD) != llvmFunctions.end()) {
+    return llvmFunctions[FD];
+  }
+  std::string name = CGM.getMangledName(FD).str();
+  std::vector<mlir::LLVM::LLVMType> types;
+  for (auto parm : FD->parameters()) {
+    types.push_back(
+        typeTranslator.translateType(getLLVMType(parm->getOriginalType())));
+  }
+
+  auto rt = typeTranslator.translateType(getLLVMType(FD->getReturnType()));
+
+  auto llvmFnType =
+      LLVM::LLVMType::getFunctionTy(rt, types,
+                                    /*isVarArg=*/FD->isVariadic());
+
+  // Insert the printf function into the body of the parent module.
```

Comment:
This function seems generic why in the comment you refer to "printf" only?

Reply:
Done